_posts/2021-05-28-why-avoid-an-orm.md

```diff
@@ -0,0 +1,231 @@
+ 
+ORMs of course provide ways to avoid n+1 problems. One way is to configure the fetching strategy for a relationship to be eager so that you can say that the employees of a department are always fetched from the database whenever the department data is fetched. But imagine the performance when there are departments with thousands of employees and your use case is only interested in the department data and not the employees.
+ 
+This can be improved by providing different fetching strategies for different use cases, but since the mapping configuration is more or less static, you can’t handle different needs easily. Well, you can (for example by using annotations/metadata to specify different fetching strategies that can be enabled one way or another), but you need to fall back to using some kind of query language where you can specify how much data you want to fetch. And since you are falling back to a query language, why not just use a query language in the first place and forget about object graphs.
+ 
+Another problem is that maybe you have pagination and you want to display only 50 employees at a time. But you can’t specify this in an object graph since all you are saying is `getEmployees()` and doing the pagination afterward. ORMs handle this by not fetching all employees in one query, but some configurable amount at a time. Again, the amount of actual queries (or physical database fetches of the result of a single query) depends on static configuration and you’d need to fall back to a query language to vary it for different use cases.
+ 
+Even though ORMs provide these means to handle the issue, the biggest problem is that you’d have to know if, when, and how to use them. And since you don’t even know when actual queries are being performed in the first place, it’s practically impossible. At least I’m not smart enough for it. Modern computers are blazingly fast, so these problems are rarely visible with any ordinary test data, and thus they creep up in production or only manifest themselves as ”well, this app feels somewhat sluggish...”.
+ 
+ 
+ 
+## Object identity
+ 
+If you have an employee, what is its identity? When are two employees the same? How do you reference a specific employee?
+ 
+In a relational database, the identity is defined by a key, but in object graphs, it would more naturally be defined by object equality. These need to be in sync, otherwise you run into weird problems, like objects disappearing when you fetch them from the database to a `Set`. 
+ 
+As far as I know, ORMs tend to handle this by forbidding object identity (don’t override object equality) and always use the database identity. Too bad you now have to live with objects that look equal but are not. A company named "Solita" isn’t equal to a company named "Solita” until you fetch them both from the database to compare their keys.
```

Comment:
This is tricky without ORMs too, comparing only names wouldn't make sense in many other contexts either.

Reply:
You are right, and I don't think I'm making much sense here.

What I think I'm trying to say is:
- Database has identity defined by the primary key.
- Objects have identity defined by the object equality.
- When using an ORM, these need to be in sync.
- When NOT using an ORM, these DO NOT need to be in sync, since when the rows are retrieved from the database they are "just data" and don't represent database rows anymore.

For example, if I have `email` table which allows duplicate addresses (maybe because the data comes from outside, or we assume that more columns will later be added with the address and thus we don't want to normalise the references), the object side must respect this identity for the ORM to work. My Email-class (even though it would only contain the address) must not override equals even though it would make sense for the code using it, like "send email to all addresses got from these lists" and we wan't to remove duplicates by just throwing the addresses inside a Set.

Would explaining something like this make more sense?

Reply:
I made some changes, does it now look any better?

Reply:
The issue is much clear now!